File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.java
class Solution {
    public int search(int[] nums, int target) {
        
        int p=pivot(nums);
      
        if(p==-1){ 
            return bis(nums,target,0,nums.length-1);
            }
        
        
        
        if(nums[p]==target)return p;
        
        else if(nums[0]<=target)return bis(nums,target,0,p-1);
        
        else{ return bis(nums,target,p+1,nums.length-1);}
       
 
        
    }
    
    int bis(int[] arr,int target,int s,int e)

        {

         while(s<=e)

        {

            int mid=s+(e-s)/2;

        if(arr[mid]==target) return mid;

        

         

        

        if(arr[mid]<target) s=mid+1;

        

        else{ e=mid-1;}

            }
        return -1;
        }
    
 
int pivot(int[] nums)
    {
    
    int s=0,e=nums.length-1;
    
    while(s<=e)
        {
        int mid=s+(e-s)/2;
        if(mid<e && nums[mid]>nums[mid+1])return mid;
   
         if(mid>s && nums[mid]<nums[mid-1])return mid-1;
        
        else if(nums[mid]<nums[s]) e=mid-1;
        
        else{
            s=mid+1;
            }
        }
    return -1;
          
        
        
        
        
        }
    

        

        

       

       

        

      

           

        

       

         

      

        

        

        

    

    

    
}

File: 0033-search-in-rotated-sorted-array/0033-search-in-rotated-sorted-array.c
int search(int* nums, int numsSize, int target) {
    int pivot = findPivot(nums, 0, numsSize-1);
    if (pivot == -1) 
        return binarySearch(nums, 0, numsSize-1, target);
    if (nums[pivot] == target) 
        return pivot;
    if (nums[0] <= target) 
        return binarySearch(nums, 0, pivot-1, target);
    return binarySearch(nums, pivot+1, numsSize-1, target);
}

int findPivot(int* nums, int low, int high) {
    if (high < low) 
        return -1;
    if (high == low) 
        return low;
    int mid = (low + high)/2;
    if (mid < high && nums[mid] > nums[mid + 1])
        return mid;
    if (mid > low && nums[mid] < nums[mid - 1])
        return (mid-1);
    if (nums[low] >= nums[mid])
        return findPivot(nums, low, mid-1);
    return findPivot(nums, mid + 1, high);
}

int binarySearch(int* nums, int low, int high, int target) {
    if (high < low)
        return -1;
    int mid = (low + high)/2;
    if (target == nums[mid])
        return mid;
    if (target > nums[mid])
        return binarySearch(nums, (mid + 1), high, target);
    return binarySearch(nums, low, (mid -1), target);
}